### PR TITLE
Add EditMessage methods

### PR DESCRIPTION
**Pull request by samuelyallop-pusher**
Adds functions for (upcoming) message editing endpoints.

Riders:
- URL encode parameters before splicing them into URLs
- Close response bodies if they exist, even if an error occurs

### CODE REVIEW

File: internal/core/core.go [hunk 1]
```diff
@@ -717,6 +720,112 @@ func (cs *coreService) DeleteMessage(ctx context.Context, options DeleteMessageO
...
+		return err
+	}
+
+	response, err := common.RequestWithUserToken(
+		cs.underlyingInstance,
+		ctx,
+		options.SenderID,
+		client.RequestOptions{
+			Method: http.MethodPut,
+			Path:   fmt.Sprintf("/rooms/%s/messages/%d", options.RoomID, options.MessageID),
```
Comment (luismfonseca) [inline internal/core/core.go]: Url encode `options.RoomID`

File: internal/core/core.go [hunk 2]
```diff
@@ -717,6 +720,112 @@ func (cs *coreService) DeleteMessage(ctx context.Context, options DeleteMessageO
...
+			Body:   requestBody,
+		})
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
```
Comment (luismfonseca) [inline internal/core/core.go]: This defer must come before the `err != nil` check.

File: internal/core/core.go [hunk 3]
```diff
@@ -717,6 +720,112 @@ func (cs *coreService) DeleteMessage(ctx context.Context, options DeleteMessageO
...
+	if options.SenderID == "" {
+		return errors.New("You must provide the ID of the user editing the message")
+	}
+
+	if options.RoomID == "" {
+		return errors.New("You must provide the ID of the room in which the message to edit belongs")
+	}
+
+	requestParts := make([]interface{}, len(options.Parts))
+	g := errgroup.Group{}
```
Comment (luismfonseca) [inline internal/core/core.go]: Good use of errgroup!

Given that have a context, we can use it in the error group `WithContext`:
```suggestion
	g, gCtx := errgroup.WithContext(ctx)
```

File: internal/core/core.go [hunk 4]
```diff
@@ -717,6 +720,112 @@ func (cs *coreService) DeleteMessage(ctx context.Context, options DeleteMessageO
...
+
+	for i, part := range options.Parts {
+		switch p := part.(type) {
+		case NewAttachmentPart:
+			i := i
+			g.Go(func() error {
+				uploadedPart, err := cs.uploadAttachment(ctx, options.SenderID, options.RoomID, p)
```
Comment (luismfonseca) [inline internal/core/core.go]:
```suggestion
				uploadedPart, err := cs.uploadAttachment(gCtx, options.SenderID, options.RoomID, p)
```

File: internal/core/core.go [hunk 5]
```diff
@@ -717,6 +720,112 @@ func (cs *coreService) DeleteMessage(ctx context.Context, options DeleteMessageO
...
+		client.RequestOptions{
+			Method: http.MethodPut,
+			Path:   fmt.Sprintf("/rooms/%s/messages/%d", options.RoomID, options.MessageID),
+			Body:   requestBody,
+		},
+	)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
```
Comment (luismfonseca) [inline internal/core/core.go]: More it before the `err != nil`. We might need to update other places if we do this everywhere...

File: internal/core/core.go [hunk 6]
```diff
@@ -717,6 +720,112 @@ func (cs *coreService) DeleteMessage(ctx context.Context, options DeleteMessageO
...
+		return err
+	}
+
+	response, err := common.RequestWithUserToken(
+		cs.underlyingInstance,
+		ctx,
+		options.SenderID,
+		client.RequestOptions{
+			Method: http.MethodPut,
+			Path:   fmt.Sprintf("/rooms/%s/messages/%d", options.RoomID, options.MessageID),
```
Comment (luismfonseca) [inline internal/core/core.go]: URL encode RoomID

File: internal/core/core.go (hunk 5 above)
Comment (samuelyallop-pusher) [reply internal/core/core.go]: We do it everywhere.

**Pull request merged**